Build error on stable releases

**Godot version:** 6a88e22423b5212fdb660d3cacad362382fbb805


**OS/device including version:** Ubuntu 17.10 64 bit


**Issue description:**
Tags 3.0.4-stable and 3.0.5-stable wont build tried both gcc and clang same errors


**Steps to reproduce:**
git checkout 3.0.4-stable or 3.0.5-stable
scons platform=x11 bits=64 tools=yes target=release_debug -j 4 (use_llvm=yes)

**OUTPUT**
scons: Reading SConscript files ...
Enabling ALSA
Enabling PulseAudio
YASM is necessary for WebM SIMD optimizations.
WebM SIMD optimizations are disabled. Check if your CPU architecture, CPU bits or platform are supported!
Checking for C header file mntent.h... (cached) yes
scons: done reading SConscript files.
scons: Building targets ...
[  3%] Compiling ==> main/default_controller_mappings.gen.cpp
[  3%] Compiling ==> main/performance.cpp
main/default_controller_mappings.gen.cpp:2:10: fatal error: default_controller_mappings.h: No such file or directory
 #include "default_controller_mappings.h"
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
[  5%] Compiling ==> main/main.cpp
[  5%] scons: *** [main/default_controller_mappings.gen.x11.opt.tools.64.o] Error 1
Compiling ==> main/input_default.cpp
scons: building terminated because of errors.

--- a/core/SCsub
+++ b/core/SCsub
@@ -146,6 +146,7 @@ env.core_sources += thirdparty_obj
 # Godot source files
 
 env.add_source_files(env.core_sources, "*.cpp")
+env.add_source_files(env.core_sources, "script_encryption_key.gen.cpp")
 
 # Certificates
 env.Depends(

--- a/core/input/SCsub
+++ b/core/input/SCsub
@@ -7,19 +7,15 @@ import input_builders
 
 # Order matters here. Higher index controller database files write on top of lower index database files.
 controller_databases = [
-    "#core/input/gamecontrollerdb.txt",
-    "#core/input/godotcontrollerdb.txt",
+    "gamecontrollerdb.txt",
+    "godotcontrollerdb.txt",
 ]
 
-env.Depends("#core/input/default_controller_mappings.gen.cpp", controller_databases)
-env.CommandNoCache(
-    "#core/input/default_controller_mappings.gen.cpp",
+gensource = env.CommandNoCache(
+    "default_controller_mappings.gen.cpp",
     controller_databases,
     env.Run(input_builders.make_default_controller_mappings, "Generating default controller mappings."),
 )
 
 env.add_source_files(env.core_sources, "*.cpp")
-
-# Don't warn about duplicate entry here, we need it registered manually for first build,
-# even if later builds will pick it up twice due to above *.cpp globbing.
-env.add_source_files(env.core_sources, "#core/input/default_controller_mappings.gen.cpp", warn_duplicates=False)
+env.add_source_files(env.core_sources, gensource)

--- a/editor/SCsub
+++ b/editor/SCsub
@@ -101,6 +101,7 @@ if env["tools"]:
     )
 
     env.add_source_files(env.editor_sources, "*.cpp")
+    env.add_source_files(env.editor_sources, "register_exporters.gen.cpp")
 
     SConscript("debugger/SCsub")
     SConscript("fileserver/SCsub")

--- a/methods.py
+++ b/methods.py
@@ -14,7 +14,7 @@ from SCons.Variables.BoolVariable import _text2bool
 from platform_methods import run_in_subprocess
 
 
-def add_source_files(self, sources, files, warn_duplicates=True):
+def add_source_files(self, sources, files):
     # Convert string to list of absolute paths (including expanding wildcard)
     if isinstance(files, (str, bytes)):
         # Keep SCons project-absolute path as they are (no wildcard support)
@@ -24,17 +24,20 @@ def add_source_files(self, sources, files, warn_duplicates=True):
                 return
             files = [files]
         else:
+            # Exclude .gen.cpp files from globbing, to avoid including obsolete ones.
+            # They should instead be added manually.
+            skip_gen_cpp = "*" in files
             dir_path = self.Dir(".").abspath
             files = sorted(glob.glob(dir_path + "/" + files))
+            if skip_gen_cpp:
+                files = [f for f in files if not f.endswith(".gen.cpp")]
 
     # Add each path as compiled Object following environment (self) configuration
     for path in files:
         obj = self.Object(path)
         if obj in sources:
-            if warn_duplicates:
-                print('WARNING: Object "{}" already included in environment sources.'.format(obj))
-            else:
-                continue
+            print('WARNING: Object "{}" already included in environment sources.'.format(obj))
+            continue
         sources.append(obj)
 
 

